### ISSUE TITLE
Make windows titles optional

### ISSUE THREAD

**Issue by TS-CUBED** [pop-os/cosmic-epoch#275]
I use some apps which are meant to not have windows titles, but the Cosmic compositor seems to force them.

One of these is Emacs, which should basically look like it's running in a terminal, but looks like this:  

![screenshot-2024-04-24-10-28-42](https://github.com/pop-os/cosmic-epoch/assets/52218856/347ae851-4e95-4533-ad24-7bc299d6f514)

I am running the pgtk version of Emacs 29.3, so it should be Wayland only. But it also is the same issue on the X-version.

Will this be a feature in the final version (disable windows titles)?

**Comment by lukejcollins** [comment]
Being able to get rid of window titles is definitely an essential feature for me - I don't find any use for them at all; they're just taking up screen real estate.

**Comment by mmstick** [comment]
We could have a field for disabling window titles in the CosmicCompConfig. Then either a toggle added to the tiling applet and/or cosmic-settings.

**Comment by TS-CUBED** [comment]
I think it would make sense

**Comment by sgued** [comment]
One more reason to have the ability to remove title bars, if you have a custom color theme, the title bar will often not match at all the application. For example on an electron app (Signal Desktop):

![image](https://github.com/user-attachments/assets/52d3bb3f-168b-48cf-be42-1b76f898870d)

**Comment by NotLebedev** [comment]
First of all thanks to all people working on cosmic, trying it out and finding it great!

Currently there are several open issues that in my opinion all are part of this problem:
1. Qt apps use csd https://github.com/pop-os/cosmic-epoch/issues/542 (and also apps like firefox, electron apps like vscode and some other apps like alacritty https://github.com/pop-os/cosmic-comp/issues/349) 
2. There is no way to remove close button (I think some issues mentioned it but can't find now)
3. There is no way to style ssds at all https://github.com/pop-os/cosmic-comp/issues/549

There is clear some interest in titlebar-less mode and appropriate style for native cosmic apps. KDE does this fine, the options to disable titlebars are a little obscure but behavior is consitent if you figure out how to do it. I think this would be a great feature and some important parts like super+click to drag and super+right click to resize are already in place so UX wise removing titlebars should be absolutely fine

**Comment by TheESN** [comment]
> One more reason to have the ability to remove title bars, if you have a custom color theme, the title bar will often not match at all the application. For example on an electron app (Signal Desktop):
> 
> ![image](https://private-user-images.githubusercontent.com/51865119/356482421-52d3bb3f-168b-48cf-be42-1b76f898870d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU0MDcyNjIsIm5iZiI6MTcyNTQwNjk2MiwicGF0aCI6Ii81MTg2NTExOS8zNTY0ODI0MjEtNTJkM2JiM2YtMTY4Yi00OGNmLWJlNDItMWI3NmY4OTg4NzBkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTAzVDIzNDI0MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE3MDRlNWU3YWE0MzU0MzQxNzlmMjI0YjkwNTkyN2E0MGE3MmUwYjYyNzU1ZDE0ZjM2NjE5NDM3NmFmZTczOWYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.VkKhussYma2PkpHtNMwakwMWidkl8yJue_ZZ6JkLm8E)

Server Side Decorations never supposed to match applications' colours. For example on windows if you enable accent colour headerbar it would look like image here. And if you disable this headerbar - you will loose window controls buttons

**Comment by sgued** [comment]
> For example on windows if you enable accent colour headerbar it would look like image here

I don't think Windows should be the benchmark on having a uniform look and feel.

> And if you disable this headerbar - you will loose window controls buttons

Yes, this feature should probably be described as "advanced" or something. But I'm migrating from sway where I already disable titlebars and only use keyboard shortcuts to manage apps.

Also note that some apps (I know of Firefox at least), can optionally show their own decorations.

**Comment by RAVENz46** [comment]
> And if you disable this headerbar - you will loose window controls buttons

Therefore, I hope https://github.com/pop-os/cosmic-applets/issues/204 is implemented at the same time.
Only for the focus-window but better than nothing.

**Comment by TheESN** [comment]
I wanted just to point out that server side decorations were never supposed to match colouring of applications since environment doesn't know what colour it is. Removing SSDs for applications that specifically asks it (COSMIC actually prefers Client Side Decorations, unless asked by application) can make some applications poorly usable since you won't be able to move them in floating mode

**Comment by NotLebedev** [comment]
> Removing SSDs for applications that specifically asks it (COSMIC actually prefers Client Side Decorations, unless asked other way by application) can make some applications poorly usable since you won't be able to move them in floating mode

Many desktops (including COSMIC!) support dragging windows with super+left click anywhere on window and resizing with super+right click anywhere near edges. This works even in floating mode. I assume that users who specifically want to disable title bars know about this feature. Additionally this functionality can be advertised in some text near opinion that disables title bars or maybe in some popup confirming the choice. I believe this not only does not degrade useability but actually improves it

**Comment by lucas75** [comment]
>> Removing SSDs for applications that specifically asks it (COSMIC actually prefers Client Side Decorations, unless asked other way by application) can make some applications poorly usable since you won't be able to move them in floating mode

> Many desktops (including COSMIC!) support dragging windows with super+left click anywhere on window and resizing with super+right click anywhere near edges. This works even in floating mode. I assume that users who specifically want to disable title bars know about this feature. Additionally this functionality can be advertised in some text near opinion that disables title bars or maybe in some popup confirming the choice. I believe this not only does not degrade useability but actually improves it

Yes! Let me share my experience:

In GNOME:
* SuperKey+RightClick displays the window menu (same as right click on TitleBar)
* SuperKey+MouseDrag moves the window
* SuperKey show the activities view (where the top bar, dash, open window list and workspace list are)
* There are extensions to hide the titlebar when the window is maximized (or merge it with the top panel);
* There are extensions to display the top panel only in the activities view;
* There are extensions to display the dash only in the activities view;
* There are configurations to make alt+tab work by window (without grouping applications).

This is my daily driver and is awesome. 
I found no windows that break, become ugly or unresponsive because of the hidden TitleBar;
Those extensions are very popular and this suggests that is a productive and desirable workflow;

It would be nice to have something like this as an option on COSMIC too.

**Comment by davxy** [comment]
I guess that for most users of i3, sway, hyprland, niri, and other minimal window managers (myself included), the ability to hide the titlebar in tiling mode is certainly an important feature.
I'd be nice to see this option into the COSMIC settings

**Comment by canadaduane** [comment]
A subtle variation that I'd find helpful, when not in tiling mode: disable the titlebar when the app is maximized, but show it when the app does not have full screen. This way, it releases some extra space to the user when switching (via key combo) between maximized / normal window size.

**Comment by kynrai** [comment]
> I guess that for most users of i3, sway, hyprland, niri, and other minimal window managers (myself included), the ability to hide the titlebar in tiling mode is certainly an important feature. I'd be nice to see this option into the COSMIC settings

I'm in that same boat. While I have no knowledge of how hard this is to implement neatly into the cosmic design language, I'm happy to test this across distros and apps. Currently  daily driving cosmic alpha 4 on fedora to test.

**Comment by diegoldnascimento** [comment]
Yup! I agree with everyone here: having an option to hide title bars is such a great feature for Cosmic DE. I can't see myself using it without ricing the UI to be as minimalistic as possible. +1

**Comment by ItsMe-TJ** [comment]
I posted a feature request, which was closed due it being a duplicate of this one. I guess I should've searched better. Oops! My bad!

Anyway, the ability to also force titlebars would be great, and not just hide them. In my original post I talked about DaVinci Resolve being the main program that imo needs a titlebar but doesn't natively have one. Window Rules is the one feature from KWin that I miss the most when I'm on other Desktop Environments. 
I would love to see COSMIC get this!

**Comment by FelipeFMA** [comment]
I think it would be great to have such an option in the tiling applet, so users can have a more tiling-WM-like experience if they prefer.

**Comment by tomengland** [comment]
Has anyone found a way to make this possible currently?  Otherwise support this feature.

**Comment by terratrax** [comment]
Came to agree with the above sentiment - users coming from existing tiling WM like hyprland will be immediately turned off by the unnecessary loss of screen real estate to titlebars. I love Cosmic and really want to make the switch but this missing feature is a showstopper for me, especially on laptops.

**Comment by rsbhalla** [comment]
Agreed. The title bar issue and the inability to set application specific float/tile options are the two biggest drawbacks when moving from i3/sway to cosmic, which otherwise is a pleasure to use.

**Comment by vstovbunsky** [comment]
Agreed with everyone here - coming from Hyprland, even on a 4k screen having a thick title bar in Blender for example is annoying. You can already close the window with SUPER+Q, move/resize it with SUPER+Mouse or tile/float with other hotkeys, making the titlebar completely optional and only useful to people who prefer mostly using mouse.
At least having an option to toggle them on/off would be fantastic.

**Comment by AKIL-blender** [comment]
I also agree with the people here: it would be good if it were possible to remove the title bar.
To move around, I only use shortcuts; it's annoying to see a bar.
I'm messing with these files to see if I can do anything, but I don't think so.
~/.config/cosmic/com.system76.CosmicSettings.WindowRules/v1

**Comment by diegoldnascimento** [comment]
Agreed with everyone here

**Comment by seb-hyland** [comment]
^ Would like to agree

**Comment by myth0genesis** [comment]
Came here to upvote this. In my opinion it's an essential feature for tiling WMs because individual windows already have so little real estate when more than one is open.

**Comment by machin3io** [comment]
Just chiming in, to confirm. This would be highly desired, and almost an essential feature for me.

**Comment by zettlrobert** [comment]
Came here to add that it would be nice to be able to disable window titles.

Currently, `wezterm` can't disable the 'titlebar' `config.window_decorations = "RESIZE"/"NONE"` is not applying.

I assume the compositor is forcing it?

**Comment by oxysoft** [comment]
This is one of the most mission critical feature in a OS. I will attempt to vibe code this soon with Claude as well as binding it to maximized state.

**Comment by ramonicv** [comment]
Coming from hyprland (like many others), would love to be able to disable titlebar (specifically an issue for me in my terminal, `wezterm`, and tui programs ran within it).

**Comment by Miguel-Escobar** [comment]
I also highly desire this feature.

**Comment by BladeBoles** [comment]
Just commenting to say I super agree with the consensus of this thread.  Hiding title bars is the only feature missing that I need to be able to adopt cosmic as my DE.  I'm coming from i3/sway and any waste of vertical space is a big deal for me.  I'm glad to see I'm not alone.  There are literally dozens of us!

**Comment by bytehoarder** [comment]
I'm also coming from `hyprland`, I spam terminals and the window title definitely adds up once you have enough terminals open. Would love to have the possibility to disable window titles!

**Comment by Justice-Caban** [comment]
I'd rather have the screen space than a titlebar, keyboard commands cover all of my needs. I'd love to have the option to disable them as well

**Comment by cramt** [comment]
I'm currently using this patch to remove it, in case anyone is building their own version from source https://github.com/cramt/nixconf/blob/main/patches/no_ssd.patch

**Comment by Dakedres** [comment]
I think one of COSMIC's unique strongsuits is how flexible it is to different changing input methods. When I'm working I definitely am more of a keyboard-centric tiling user, but when I'm using the computer more leisurely I find myself becoming a mouse user, and then I need the titlebars. An option that would also be interesting to explore is hiding the titlebar until we hover over the top edge of the window. Best of both worlds IMO (and a great compositor flex).

**Comment by eliot-akira** [comment]
I'd love to see this feature also, a way to hide window title bars on all applications. It's one of the last few details I'm still solving after upgrade to Cosmic.

There is some progress in this direction, thanks to @cramt.

- pop-os/cosmic-comp#1385
- pop-os/libcosmic#865

The latter PR is currently blocking it, but it needs someone with authority/ownership to make a decision on how the feature should be designed. (I think it's better as an on/off setting instead of custom height, which seems to be the disagreement there.)

**Comment by cramt** [comment]
yeah sorry y'all i havnt had the time to maintain these branches, thats why i've just swapped to using a simple patch file in my setup

**Comment by eliot-akira** [comment]
Thanks for sharing a [link to the patch](https://github.com/cramt/nixconf/blob/main/patches/no_ssd.patch) too, I'm now learning how to apply it. Found some good info here with more tips:

- [NixOS Wiki: COSMIC Desktop Environment](https://wiki.nixos.org/wiki/COSMIC)
- https://github.com/NixOS/nixpkgs/issues/259641

**Comment by ElnuDev** [comment]
@eliot-akira If you figure out how to apply the patch, please let us know! Usually in the NixOS options there's a `package` option that lets you override the package used (e.g. `services.xserver.windowManager.i3.package` for i3) but as far as I could tell, [there aren't any available similar options for COSMIC](https://search.nixos.org/options?channel=unstable&query=cosmic), so I wasn't really sure how to do it. I suspect the only way would be to make a nixpkgs overlay and override the source there.

**Comment by threddast** [comment]
@ElnuDev this is all you need:
```

  nixpkgs.overlays = [
    (_: prev: {
      cosmic-comp = prev.cosmic-comp.overrideAttrs (old: {
        patches = (old.patches or [ ]) ++ [ ./no_ssd.patch ];
      });
    })
  ];

```

**Comment by ElnuDev** [comment]
@threddast Thank you, this worked perfectly!

If anybody is wondering, this doesn't apply to Firefox, but it's very easy to remove the title bar within Firefox. Right click on toolbar > Customize Toolbar... > at bottom of screen, uncheck Title Bar

**Comment by jcornuz** [comment]
Maybe of use for others: attached is a patch that optionally removes the Cosmic SSD on maximized and / or tiled windows.

It adds the required logic in the windows display and creates two new options in cosmicComp / appearance_settings: 
- ssd_tiled_windows 
- ssd_maximized_windows 

If set to false, SSD are hidden on tiled / maximzed windows (default is true: keep the SSD on all the time)

Disclaimer(s): this is vibecoded (but very simple); it works for me (hopefully for others); it removes only cosmic-drawn window titlebar, not GTK or QT ones.

[option_for_no_ssd_on_maximized_or_tiled_windows.patch](https://github.com/user-attachments/files/28415168/option_for_no_ssd_on_maximized_or_tiled_windows.patch)

**Comment by ecoezen** [comment]
> I use some apps which are meant to not have windows titles, but the Cosmic compositor seems to force them.
> 
> One of these is Emacs, which should basically look like it's running in a terminal, but looks like this:
> 
> ![screenshot-2024-04-24-10-28-42](https://github.com/pop-os/cosmic-epoch/assets/52218856/347ae851-4e95-4533-ad24-7bc299d6f514)
> 
> (Edit: Oops, looks like the screenshot tool doesn't like rotated screens. Will open another issue on that one.)
> 
> I am running the pgtk version of Emacs 29.3, so it should be Wayland only. But it also is the same issue on the X-version.
> 
> Will this be a feature in the final version (disable windows titles)?

Emacs is the screenshot in the original post so figured I'd share what I found for it specifically. No compositor patch needed, fixable from the Emacs config.

default-frame-alist only applies when a frame is created. Plain emacs's first frame gets created before your init files run, so undecorated isn't set yet. GTK3 sees decorated = TRUE and requests Client from the KDE decoration protocol, so no titlebar.

Any frame created after init has already loaded (emacsclient -c, make-frame) already has undecorated in default-frame-alist at creation time. GTK3 requests Server instead, and cosmic-comp goes along with it. Same config, different frame creation timing, different result.

Fix: don't put undecorated in default-frame-alist. Set it after the frame exists instead.

```elisp
;; don't do this, this is what causes the inconsistency:
;; (add-to-list 'default-frame-alist '(undecorated . t))

(defun my/undecorate-frame (&optional frame)
  (set-frame-parameter (or frame (selected-frame)) 'undecorated t))

(add-hook 'emacs-startup-hook #'my/undecorate-frame)
(add-hook 'after-make-frame-functions #'my/undecorate-frame)
(add-hook 'server-after-make-frame-hook #'my/undecorate-frame)
```

Added some debug logging to KdeDecorationHandler::request_mode to check this. With the hooks in place, every frame type only sends one Client request, at creation, and nothing after. Looks like GTK negotiates decoration mode once at surface creation and doesn't redo it when decorated changes later on a frame that's already mapped.

One thing I noticed but aren't sure is actually a problem: request_mode accepts whatever mode the client sends with no check against the previously recorded mode. Didn't affect us since Emacs never re-requests after the fix, but someone who knows this code and the protocol better might want to check it.

**Comment by BerriJ** [comment]
> I'm currently using this patch to remove it, in case anyone is building their own version from source https://github.com/cramt/nixconf/blob/main/patches/no_ssd.patch

This patch somehow stopped working for me. I'm not quite sure why. Currently I'm using cosmic 1.2.0.
I'll update this comment once I have tested with cosmic 1.4.0.

**Comment by cramt** [comment]
For anyone following: pop-os/cosmic-comp#2666 adds `appearance_settings.show_titlebars` to the compositor config — set it to `false` and cosmic stops drawing the server-side titlebar for windows that don't decorate themselves. (Replaces the older pop-os/cosmic-comp#1385 + pop-os/libcosmic#865 pair, which took the configurable-SSD-height route that maintainers rejected.)
